Revise start_fact_cache and finish_fact_cache to use JSON file

djyasin opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
This proposes a revision to our fact caching mechanics. Description of the change:

- for start_fact_cache: Instead of returning the tuple of 2 items (host-list, write-time), at the end of the method (after the loop completes), we write a JSON format file with the host list inside of it
- This JSON format file will be in the "artifacts" directory, but not inside of the host_facts directory
- for finish_fact_cache: Instead of accepting (host-list, write-time) as arguments, it obtains:
- host-list from a load of the JSON format file
- write-time as the modtime of that same file
- Corresponding change is that the "destination" function arguments is changed from the host_facts directory to the artifacts directory so that we can write and read this extra file.

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API


##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 24.6.2
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

djyasin committed:
Revise start_fact_cache and finish_fact_cache to use JSON file with host list inside it

Remove list from line that looks up host objects and add iterator

Update functionality of finish_fact_cache

Remove reference to last_modified

Remove unused import of time

Attempt to get test_finish_job_fact_cache_clear working

Remove unused self calls

Add back test_finish_job_fact_cache_with_bad_data

Update test_finish_job_fact_cache_clear

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -67,54 +73,74 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
+    facts_write_time = os.path.getmtime(summary_path)
+
+    # Lookup Host objects
+    hosts_cached = list(Host.objects.filter(name__in=host_names, inventory_id=inventory_id).order_by('id'))
```
This materializes all the ansible facts for all the hosts in the inventory into memory.

With the current structure, I believe you can avoid that by simply removing the `list(` and adding an `.iterator()` to the queryset. It won't load the queryset into memory right away. And the iterator allows garbage collecting hosts that have gone out of scope in the loop.

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -67,54 +73,74 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
+        return
+
+    host_names = summary.get('hosts_cached', [])
+    facts_write_time = os.path.getmtime(summary_path)
```
This should work the same as before. My idea to solve the UTC problem was that we might use `os.path.getmtime` of _this_ file at `summary_path`. That way, if the remove node rewrites the modtime of the files, it does so consistently for the artifact files as well as this summary file.

djyasin replied:
I believe I've addressed this comment now!

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -34,31 +35,36 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
-    if last_filepath_written:
-        return os.path.getmtime(last_filepath_written), hosts_cached
-
-    return None, hosts_cached
```
From the failures, it looks like other locations are still expecting a return of a tuple. So your change just needs to be propogated.

I fully agree with changing the return type of this. It won't be expected to return anything after this. That change just needs to be propagated to the callers of the method, and any variables saving the return values need to be deleted.

djyasin replied:
Thank you so much!!

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -67,54 +73,73 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
-    hosts_cached = sorted((h for h in hosts_cached if h.id is not None), key=lambda h: h.id)
+    summary_path = os.path.join(destination, 'host_cache_summary.json')
+    if not os.path.exists(summary_path):
+        system_tracking_logger.error(f'Missing summary file at {summary_path}')
```
I looked at your test failures, and was awfully perplexed. It seems that what is making this very difficult to debut is the use of `system_tracking_logger` as opposed to `logger`. The former will present the logs in normal logs. I would avoid the former except for the very narrow cases where it is already being used. Because of this, we're not getting legitimate output from tests, which should show this log is being emitted, which is likely the current problem you're on.

djyasin replied:
I believe this has been addressed!

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -34,31 +35,36 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
-    return None, hosts_cached
+    # Write summary file to artifacts dir
+    if inventory_id is not None:
+        artifact_dir = os.path.join('artifacts', f'inventory_{inventory_id}')
```
Artifacts needs to be relative to the job private_data_dir. This is nonsense path.

This file should go at private_data_dir/artifacts/42/host_cache_summary.json, ideally.

djyasin replied:
I think I've adjusted this correctly! Please let me know if this isn't what you meant! [0e6ca8667ae9dddbbd253a4dd6dc5aaa43ea8783](https://github.com/ansible/awx/pull/15970/commits/0e6ca8667ae9dddbbd253a4dd6dc5aaa43ea8783)

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -25,7 +25,8 @@
 def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=None):
```
Right now, `destination` is passed into this method which is like: private_data_dir/artifacts/42/fact_cache/

However, with this change, some stuff needs to be written at the folder starting at: private_data_dir/artifacts/42/

So I would start trying to be way more explicit here. Instead of `destination`, take `artifacts_dir`. And then for some parts of the code you need the fact_cache dir, so best to save that to a descriptive var like `fact_cache_dir = os.path.join(artifacts_dir, 'fact_cache')` when you come to that point.

Implication is that the _caller_ need to change, and when that happens, it's far better to pass it as a kwarg, like `artifacts_dir=`, because reading the task code is very hard as it is, and making the method signature clear helps.

djyasin replied:
I believe I have accomplished this as well now!

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -67,54 +73,73 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
+    hosts_cached = Host.objects.filter(name__in=host_names, inventory_id=inventory_id).order_by('id').iterator()
 
     hosts_to_update = []
+    host_facts_dir = os.path.join(settings.ANSIBLE_FACT_CACHE, str(inventory_id))
```
Here also, needs directory treatment relative to artifacts dir. `settings.ANSIBLE_FACT_CACHE` isn't a thing to my knowledge.

djyasin committed:
Revise artifacts path to be relative to the job private_data_dir

Update dirs to be more specific

djyasin committed:
Update calls to start_fact_cache and finish_fact_cache to agree with new reference to artifacts_dir

Update tests to agree with new dir structure in start and finish_fact_cache

Resolve log_data issue

Add back code comments removed in error

Remove inventory_id=inventory_id in line 100 to resolve test issue

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -67,43 +75,65 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
+                except ValueError:
+                    continue
+
+                host.ansible_facts = ansible_facts
```
I'm only slightly concerned that `except ValueError` was intended to catch an error in this line. Specifically, that `ansible_facts` might contain a type not compatible with the field. Yet, this would have to be a JSON type. So... like an integer or a dictionary might raise a ValueError? If that does happen, then we would prefer to continue the loop as opposed to error the method.

But I don't feel strongly about this at the current moment, because if it's a concern, then it should have a test. However, as a code-style point, I think the exception is unlikely to be meaningful in its current location.

AlanCoding reviewed (approved):
Look great, nice work.

the dev-env tests passing are a real green flag for this. I don't believe this institutes any changes expectations for behavior, so debugging tests should be straightforward.

djyasin committed:
Prevents unnecessary updates to ansible_facts_modified, fixing timestamp-related test failures.

djulich commented on awx/main/tasks/facts.py:
```diff
@@ -67,34 +75,58 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
     log_data['cleared_ct'] = 0
+    # The summary file is directly inside the artifacts dir
+    summary_path = os.path.join(artifacts_dir, 'host_cache_summary.json')
+    if not os.path.exists(summary_path):
```
Catching the missing file error here seems redundant, because it would be caught by `except (..., OSError)` below (line 94) anyway. But on the other hand, we get a more specific error message with this extra clause.

djulich reviewed (approved):
Code looks good to me!
